docs/content/ActionList.mdx

```diff
@@ -0,0 +1,67 @@
+---
+title: ActionList
+---
+
+An `ActionList` is a list of items which can be activated or selected. `ActionList` is the base component for many of our menu-type components, including `DropdownMenu` and `ActionMenu`.
+
+## Minimal example
+
+```jsx live
+<ActionList
+  items={[
+    {text: 'New file'},
+    ActionList.Divider,
+    {text: 'Copy link'},
+    {text: 'Edit file'},
+    {text: 'Delete file', variant: 'danger'}
+  ]}
+/>
+```
+
+## Example with grouped items
+
+```jsx live
+<ActionList
+  groupMetadata={[
+    {groupId: '0'},
+    {groupId: '1', header: {title: 'Live query', variant: 'subtle'}},
+    {groupId: '2', header: {title: 'Layout', variant: 'subtle'}},
+    {groupId: '3'},
+    {groupId: '4'}
+  ]}
+  items={[
```

Comment:
Having read @T-Hugs's [doc](https://github.com/primer/components/pull/1151), I'd be curious to hear if we've tried implementing this component with subcomponents. 

I recognize that I don't have a ton of context here, and that most of my relevant experience is on the Rails side, but several existing Primer React components use subcomponents, such as [TabNav](https://primer.style/components/TabNav), and I'm wondering if it might be a better long-term option here.

The reason I ask is that it _feels_ like we are passing a lot of information to ActionList about its children, effectively having it conceptually cover two different levels of hierarchy (parent and children). While it is not onerous now, I can imagine a future where we need to double the number of options for child items, leading to a significantly larger API for this single component.

Instead, I could see us having:

```jsx
<ActionList groups={[
     {groupId: '0'},
     {groupId: '1', header: {title: 'Live query', variant: 'subtle'}},
     {groupId: '2', header: {title: 'Layout', variant: 'subtle'}},
     {groupId: '3'},
     {groupId: '4'}
   ]}>
  <ActionList.Item leadingVisual={TypographyIcon} text="Rename" groupId="0"/>
  // etc
```

By splitting the API across the two levels of hierarchy, we get to divide the number of properties between the parent and children. Having seen the breadth and depth of @vdepizzol's ideas for overlays, I expect there to be expansion in both dimensions in the future 😄

Reply:
From @joelhawksley in https://github.com/primer/components/pull/1078#discussion_r606411161:
> Having read @T-Hugs's [doc](https://github.com/primer/components/pull/1151), I'd be curious to hear if we've tried implementing this component with subcomponents.

With a subcomponent approach, can `Item`s be associated with `Group`s without resorting to `React.Children.map`? (cf. [“The `React.Children` anti-pattern”](https://github.com/primer/components/blob/c727a03fd7a933814506d81bedb71535a0e3b67a/contributor-docs/component-contents-api-patterns.md#the-reactchildren-anti-pattern))

Reply:
(You’re asking about coupling between parent and child types (“it feels like we are passing a lot of information to ActionList about its children, effectively having it conceptually cover two different levels of hierarchy (parent and children)”), but as an aside:) If syntax was also a concern, the `ActionList` API (as implemented in this PR) supports JSX in addition to JSON:

```tsx
<ActionList
  groupMetadata={[
    {groupId: '0'},
    {groupId: '1', header: {title: 'Live query', variant: 'subtle'}},
    {groupId: '2', header: {title: 'Layout', variant: 'subtle'}},
    {groupId: '3'},
    {groupId: '4'}
  ]}
  items={[
    {groupId: '0', renderItem: p => <ActionList.Item leadingVisual={TypographyIcon} text="Rename" {...p} />},
    {groupId: '0', renderItem: p => <ActionList.Item leadingVisual={VersionsIcon} text="Duplicate" {...p} />},
    {groupId: '1', renderItem: p => <ActionList.Item leadingVisual={SearchIcon} text="repo:github/memex,github/github" {...p} />},
    {groupId: '2', renderItem: p => <ActionList.Item leadingVisual={NoteIcon} text="Table" description="Information-dense table optimized for operations across teams" descriptionVariant="block" {...p} />},
    {groupId: '2', renderItem: p => <ActionList.Item leadingVisual={NoteIcon} text="Board" description="Kanban-style board focused on visual states" descriptionVariant="block" {...p} />},
    {groupId: '3', renderItem: p => <ActionList.Item leadingVisual={FilterIcon} text="Save sort and filters to current view" {...p} />},
    {groupId: '3', renderItem: p => <ActionList.Item leadingVisual={FilterIcon} text="Save sort and filters to new view" {...p} />},
    {groupId: '4', renderItem: p => <ActionList.Item leadingVisual={GearIcon} text="View settings" {...p} />}
  ]}
/>
```

Reply:
> With a subcomponent approach, can Items be associated with Groups without resorting to React.Children.map?

I think I might have missed a level of abstraction here! Would this work?

```jsx
<ActionList />
  <ActionList.Group title="'Live query" variant="subtle"} />
    <ActionList.GroupItem leadingVisual={TypographyIcon} text="Rename"/>
   // etc
```

Reply:
Personally I much prefer the data API approach here:

1. The list owns grouping, absolving the consumer from this responsibility.
2. Limits the ways it can be rendered without an explicit escape hatch (`renderItem`) while simultaneously providing a strongly-typed API.
3. Easier to support things like loading states and extensibility (`items` as a Promise).
4. Imagine a `Dropdown` component that uses `ActionList`. The `Dropdown` would need a say in the the items rendered by `ActionList` (e.g. the leading visual to show a checkmark, an `onClick` handler). This is easy if the `Dropdown` is passed items as data, which then get modified and forwarded to `ActionList`. But if your items come as children, you have to do a complicated traversal through the children with `React.Children`, using `cloneElement` to effect the right behaviors.

In the interest of tempo, I would like to get this PR merged so we can unblock other components.

Reply:
Thanks for enumerating some of the benefits of an items-as-data approach, @T-Hugs. 👍 

> In the interest of tempo, I would like to get this PR merged so we can unblock other components.

Sounds good 👍 I’m merging this PR, because doing so doesn’t preclude us from supporting other component APIs in the future: `ActionList` _could_ be updated to coalesce its `React.Children` into an `items`-like array side-by-side with this PR’s `items` prop. If whether `ActionList` _should_ do so is a topic of continued interest, merging is non-disruptive.

Merging unblocks #1152 and #1135 and allows us to ship-to-learn about the ergonomics and DX of this component API outside of contrived Storybook stories, e.g. in situations where the list of items is not available when the component is initially loaded (as anticipated by @t-hugs in [“3.” above](https://github.com/primer/components/pull/1078#discussion_r607100270)).

Reply:
Related: https://github.com/github/primer/discussions/143